src/managers/backup.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for all backup/restore related tasks."""
+
+import logging
+from datetime import datetime
+
+import boto3
+from botocore.client import Config
+from botocore.exceptions import ClientError
+from mypy_boto3_s3.service_resource import Bucket
+
+from common.client import EtcdClient
+from common.exceptions import EtcdBackupError
+from core.cluster import ClusterState
+from core.workload import WorkloadBase
+from literals import BACKUP_FILE_PATH, BACKUP_ID_FORMAT, INTERNAL_USER
+
+logger = logging.getLogger(__name__)
+
+
+class BackupManager:
+    """Manage everything related to backup and restore."""
+
+    def __init__(self, state: ClusterState, workload: WorkloadBase):
+        self.state = state
+        self.workload = workload
+        self.admin_user = INTERNAL_USER
+        self.admin_password = self.state.cluster.internal_user_credentials.get(INTERNAL_USER, "")
+
+    def get_bucket_resource(self, s3_parameters: dict[str, str]) -> Bucket:
+        """Get the Bucket resource from the s3 connection.
+
+        Returns:
+            Bucket: the s3 bucket for uploading/downloading backups
+        """
+        s3_resource = boto3.resource(
+            "s3",
+            region_name=s3_parameters.get("region"),
+            endpoint_url=s3_parameters["endpoint"],
+            aws_access_key_id=s3_parameters["access-key"],
+            aws_secret_access_key=s3_parameters["secret-key"],
+            config=Config(
+                # https://github.com/boto/boto3/issues/4400#issuecomment-2600742103
+                request_checksum_calculation="when_required",
+                response_checksum_validation="when_required",
+            ),
+            verify=self.workload.paths.tls.backup_ca
+            if s3_parameters.get("tls-ca-chain")
+            else True,
+        )
+
+        return s3_resource.Bucket(s3_parameters["bucket"])
+
+    def create_bucket(self, s3_parameters: dict[str, str]) -> None:
+        """Create bucket if it does not exist yet."""
+        region = s3_parameters.get("region")
+        bucket = self.get_bucket_resource(s3_parameters)
+
+        try:
+            if region:
+                bucket.create(CreateBucketConfiguration={"LocationConstraint": region})
+            else:
+                bucket.create()
+            bucket.wait_until_exists()
+        except ClientError as e:
+            if (
+                # AWS returns these if the bucket was already created
+                "BucketAlreadyOwnedByYou" in e.args[0]
+                or "BucketAlreadyExists" in e.args[0]
+                # GCP returns this if the bucket was already created
+                or "BucketNameUnavailable" in e.args[0]
+            ):
+                logger.info(f"Using existing bucket {s3_parameters['bucket']}")
+                return
+            else:
+                # todo: do we want to raise to make the user aware of the error?
```

Comment:
yes - we should definitely return it to the user in the action: `Backup failed. Check error logs for more details.`